iOS Client/Store/MBArticlesStore.swift

```diff
@@ -17,184 +17,172 @@ class MBArticlesStore: NSObject {
+        performDownload(clientFunction: client.getCategoriesWithCompletion, managedObjectContext: managedObjectContext, deserializeFunc: MBCategory.deserialize, completion: completion)
     }
     
-    private func downloadArticles(persistentContainer: NSPersistentContainer, completion: @escaping (Bool?, Error?) -> Void) {
-        performDownload(clientFunction: client.getArticlesWithCompletion, persistentContainer: persistentContainer, deserializeFunc: MBArticle.deserialize, completion: completion)
+    private func downloadArticles(managedObjectContext: NSManagedObjectContext, completion: @escaping (Bool?, Error?) -> Void) {
+        performDownload(clientFunction: client.getArticlesWithCompletion, managedObjectContext: managedObjectContext, deserializeFunc: MBArticle.deserialize, completion: completion)
     }
     
     // An internal helper function to perform a download
-    private func performDownload(clientFunction: (@escaping ([Data], Error?) -> Void) -> (),
-                                 persistentContainer: NSPersistentContainer,
+    private func performDownload(clientFunction: (@escaping ([Data], Error?) -> Void) -> Void,
+                                 managedObjectContext: NSManagedObjectContext,
                                  deserializeFunc: @escaping (NSDictionary, NSManagedObjectContext) throws -> Bool,
                                  completion: @escaping (Bool?, Error?) -> Void) {
         
-        var newDataFlag: Bool = false
         clientFunction { (data: [Data], err: Error?) in
             if let clientErr = err {
                 completion(false, clientErr)
                 return
             }
             
-            // This serialQueue enables us to be certain only to invoke the completion handler once
-            // It also guards us from race conditions when writing to the numHandled variable
-            let serialQueue = DispatchQueue(label: "syncpoint")
-            var numHandled = 0
-            var hasReturned = false
-            
-            // data is an array of Data, where each datum is a serialized array
-            // thus, think of data as an array of arrays of objects
+            var isNewData = false
+            var caughtError: Error? = nil
+            // data is an array of Data, where each datum is a serialized array representing a page of results
+            // thus, think of data as an array of results pages
             for jsonData in data {
                 do {
-                    try self.downloadModelsHandler(persistentContainer: persistentContainer,
+                    isNewData = try self.downloadModelsHandler(managedObjectContext: managedObjectContext,
                                                    data: jsonData,
-                                                   deserializeFunc: deserializeFunc,
-                                                   completion: { (isNewData: Bool?, err: Error?) in
-                        serialQueue.async {
```

Comment:
What happened to the serial queue? we don't need this anymore?

Reply:
It's a subtle thing, but `downloadModelsHandler` used to be asynchronous, so we'd pass it a completion block. Internally, it would use the persistentContainer's performBackgroundTask, which was `async`. The serialQueue was basically the sync point as these background tasks came back. 

After some more thought, I expect that these background tasks were being queued on the same managed object context, meaning they would happen in a serial order and we wouldn't really need this serialQueue to guard against threads colliding into each other. But that was my fear at the time.

Anyway, now downloadModelsHandler is synchronous b/c we are passing in the managedObjectContext. The cool thing I learned is that you don't have to worry so much about threads if you just use the block-based APIs `perform` and `performAndWait` for NSManagedObjectContext. They guarantee that the work happens in the queue designated for the context. 

From the docs:

```
perform(_:) and performAndWait(_:) ensure the block operations are executed on the queue specified for the context. The perform(_:) method returns immediately and the context executes the block methods on its own thread. With the performAndWait(_:) method, the context still executes the block methods on its own thread, but the method doesn’t return until the block is executed.
```

This allowed me to make `downloadModelsHandler` sync, and greatly simplified this function that calls downloadModelsHandler for every page